post.md

```diff
@@ -0,0 +1,427 @@
+  const { status, newId } = formState;
+  const { isOpen, open, close } = useModal();
+
+  useEffect(() => {
+    if (status === "success") {
+      setId(newId);
+      close();
+    }
+  }, [status, newId, close]);
+
+  return (
+    <>
+      <div className="flex flex-col h-full text-center">
+        <div className="h-full bg-white lg:mt-px lg:py-5 px-8 dark:bg-slate-900">
+          <span className="mt-7 font-bold text-xl text-gray-600 dark:text-gray-200">
+            Dialog confirm
+          </span>
+        </div>
+
+        <div className="bg-white flex justify-center lg:mt-px pt-7 dark:bg-slate-900">
+          {id}
+        </div>
+
+        <div className="bg-white py-8 px-8 dark:bg-slate-900">
+          <Button text="Click Me!" onClick={open} />
+        </div>
+      </div>
+
+      <Modal
+        isOpen={isOpen}
+        close={close}
+        title="Confirm"
+        content="Are you sure?"
+      >
+        <form action={formAction} className="inline-block">
+          <input type="hidden" name="type" value="dialog-confirm" />
+          <input type="hidden" name="id" value={id} />
+
+          <Button text="Cancel" onClick={close} className="mr-2" />
+          <ServerActionButton text="Ok" />
+        </form>
+      </Modal>
+    </>
+  );
+};
+```
+
+There is a lot going on here, so let's break it down.
+
+First we need to import the server action we want to use. In this case we are using `deleteSomethingWithResponse` which returns a random id.
+
+```tsx
+import { deleteSomethingWithResponse } from "@/app/server-actions";
+```
+
+The server action implementation:
+
+```ts
+"use server";
+
+export async function deleteSomethingWithResponse(
+  _prevState: any,
+  params: FormData
+) {
+  const id = params.get("id");
+  const type = params.get("type");
+
+  console.log(`[${type}] Deleting ${id}...`);
+
+  // Simulate a long-running operation
+  await new Promise((resolve) => setTimeout(resolve, 2000));
+
+  const newId = crypto.getRandomValues(new Uint32Array(1))[0].toString();
+
+  return { status: "success", message: `Deleted ${id}!`, newId };
+}
+```
+
+Note the return value, this will be assigned to `formState` in the client component by `useFormState` hook once the submission is complete. It contains the `status` of the server action, a `message` and a random `newId` that we are going to render in the UI.
+
+We need to import `useFormState` from `react-dom`:
+
+```tsx
+// @ts-ignore
+import { experimental_useFormState as useFormState } from "react-dom";
+```
+
+We need to pass our server action to the `useFormState` hook:
+
+```tsx
+const [formState, formAction] = useFormState(deleteSomethingWithResponse, {});
+```
+
+The `formAction` function is used as the form's action attribute instead of using the server action directly. The `formState` object contains the response from the server, we are going to use the `status` to close the dialog and `newId` to update the UI.
```

Comment:
Or, "The `formState` object contains the response from the server. We will then use the `status` in that object to close the dialog, and `newId` to update the UI.